### PR TITLE
update for v8

### PR DESCRIPTION
**Pull request by mark-prins**
The structure of the data response has changed in v8 of grafana.
Rather than rewriting everything, I've mapped the new to the old and kept the excel write methods the same.
Small offroad to add debugging support, using delve. Limited success, tbh, but I've kept the changes in.

Also fixed the problem of the URL having a trailing slash, and the config page not loading.

### CODE REVIEW

File: backend/pkg/api/query.go [hunk 1]
```diff
@@ -59,7 +59,17 @@ type QueryResponse struct {
...
+				Schema struct {
+					Fields []struct {
+						Name string `json:"name"`
+						Type string `json:"type"`
+					} `json:"fields"`
+				} `json:"schema"`
+				Data struct {
+					Values [][]interface{} `json:"values"`
+				}
+			} `json:"frames"`
```
Comment (joshxg) [inline backend/pkg/api/query.go:72]: Oh ok thanks grafana

File: backend/pkg/api/query.go [hunk 2]
```diff
@@ -85,17 +95,37 @@ func NewQueryResponse(response *http.Response) (*QueryResponse, error) {
...
+			for rownum, _ := range rows {
+				row := make([]interface{}, columnCount)
+				for colnum, value := range values {
+					row[colnum] = value[rownum]
+				}
+				rows[rownum] = row
+			}
+
+			return rows
+		}
```
Comment (joshxg) [inline backend/pkg/api/query.go:112]: whats their problem

File: backend/pkg/api/users.go [hunk 3]
```diff
@@ -13,7 +13,7 @@ type User struct {
 }
 
 func GetEmails(authConfig auth.AuthConfig, userIDs []string, datasourceID int) ([]string, error) {
-	url := authConfig.AuthURL() + "/api/tsdb/query"
+	url := authConfig.AuthURL() + "/api/ds/query"
```
Comment (joshxg) [inline backend/pkg/api/users.go:16]: 😠

File: backend/Magefile.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ import (
 // See https://github.com/grafana/grafana-plugin-sdk-go/blob/002de0b2b925ac930952dc380682100dee27b439/build/common.go#L108.
 var cb build.BeforeBuildCallback = func(cfg build.Config) (build.Config, error) {
 	cfg.EnableCGo = true
+	// cfg.EnableDebug = true
```
Comment (mark-prins) [inline backend/Magefile.go:14]: needs to be enabled in order to build without optimisation flags for a 'better debugging experience'

File: backend/debug.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+Read more at https://www.kernel.org/doc/Documentation/security/Yama.txt"
+    read -p "Set ptrace_scope to 0? y/N (default N)" set_ptrace_input
+    if [ "$set_ptrace_input" == "y" ] || [ "$set_ptrace_input" == "Y" ]; then
+      echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope
+    fi
+  fi
+fi
+PLUGIN_PID=`pgrep ${PLUGIN_NAME}`
+IFS=- read -ra PLUGIN_PIDS <<< "$PLUGIN_PID"
+dlv attach ${PLUGIN_PID/${PLUGIN_PIDS[0]}} --headless --listen=:${PORT} --api-version 2 --log
```
Comment (mark-prins) [inline backend/debug.sh:21]: ok, yes, this is pretty hokey. For me, there are two processes spawned and attaching the first fails, so I'm only attaching to the second here.

File: backend/package.json [hunk 6]
```diff
@@ -8,8 +8,8 @@
     "test": "grafana-toolkit plugin:test",
     "dev": "grafana-toolkit plugin:dev",
     "watch": "grafana-toolkit plugin:dev --watch",
-    "sign": "rm ./dist/.DS_Store & grafana-toolkit plugin:sign --signatureType private --rootUrls https://demo-board.msupply.org:3000,https://ci.msupply.org:3000,https://liberia.msupply.org:3000,https://tonga-vax.msupply.org:3000,https://timor.msupply.org:3000",
-    "go": "go run mage.go && cp -r ./dist/* ~/GitHub/msupply-dashboard/custom-plugins/msupply-dashboard-app/backend/"
+    "sign": "rm -rf ./dist/.DS_Store & grafana-toolkit plugin:sign --signatureType private --rootUrls https://demo-board.msupply.org:3000,https://ci.msupply.org:3000,https://liberia.msupply.org:3000,https://tonga-vax.msupply.org:3000,https://timor.msupply.org:3000,http://localhost:3000,https://vanuatu-vax.msupply.org:3000,https://tokelau.msupply.org:3000,https://tonga.msupply.org:3000,https://who-demo.msupply.org:3000",
+    "go-build": "go run mage.go && yarn sign"
```
Comment (mark-prins) [inline backend/package.json:12]: signing being required in v8 so I've added that in to save a few seconds of typing.

File: backend/pkg/api/panel.go [hunk 7]
```diff
@@ -152,7 +152,7 @@ func (panel *TablePanel) GetData(authConfig auth.AuthConfig) error {
 		return err
 	}
 
-	url := authConfig.AuthURL() + "/api/tsdb/query"
+	url := authConfig.AuthURL() + "/api/ds/query"
```
Comment (mark-prins) [inline backend/pkg/api/panel.go:155]: yes, the URL has changed. Should really be using the backend query helpers I guess.

File: backend/pkg/auth/auth_config.go [hunk 8]
```diff
@@ -28,7 +28,13 @@ func (config AuthConfig) AuthString() string {
 }
 
 func (config AuthConfig) AuthURL() string {
-	return config.InjectAuthString()
+	authUrl := config.InjectAuthString()
```
Comment (mark-prins) [inline backend/pkg/auth/auth_config.go:31]: there's a bug, where if you append a slash to the URL the whole thing doesn't work.
This is slightly better?

File: frontend/src/components/Settings/ConfigurationForm.tsx [hunk 9]
```diff
@@ -53,7 +53,7 @@ export const ConfigurationForm: FC<FormProps> = ({ formValues, onSubmit }) => {
                 inputName="grafanaUsername"
                 invalid={!!errors.grafanaUsername}
                 errorMessage={errors.grafanaUsername?.message ?? ''}
-                register={() => register({ required: intl.get('required') })}
+                register={() => register('username', { required: intl.get('required') })}
```
Comment (mark-prins) [inline frontend/src/components/Settings/ConfigurationForm.tsx:56]: For me, the plugin config page doesn't load without this.

File: backend/debug.sh (hunk 5 above)
Comment (joshxg) [reply backend/debug.sh:21]: Yeh i'm not going to pretend to understand this, sorry. But.. cool!

LGTM

File: backend/Magefile.go (hunk 4 above)
Comment (joshxg) [reply backend/Magefile.go:14]: 😁

File: backend/pkg/api/panel.go (hunk 7 above)
Comment (joshxg) [reply backend/pkg/api/panel.go:155]: :(

if I remember the sdk didn't provide anyway to do this... quick look at activity in the sdk repo and it seems they might added this functionality with https://github.com/grafana/grafana-plugin-sdk-go/pull/328 

but not sure..

😢

File: frontend/src/components/Settings/ConfigurationForm.tsx (hunk 9 above)
Comment (joshxg) [reply frontend/src/components/Settings/ConfigurationForm.tsx:56]: I suppose grafana is injecting in a newer version of grafana/ui ?

https://github.com/grafana/grafana/commit/3b515e650c55f0125020660d8aa22811f756b0d3#diff-f4a60751706bfa100bd13b2965a028d8e0e4db8a3592e0e701f26696afc3e5f1R267

**Review by joshxg — APPROVED** [review]
Awesome. sorry you had to deal with that..

**Comment by mark-prins** [comment]
> if I remember the sdk didn't provide anyway to do this

Yeah, that's what I remember too. When poking around with this I saw a method that looked like it did the business, but can't see it now. Also didn't want to refactor things, even if it is neater. Maybe next time.

> I suppose grafana is injecting in a newer version of grafana/ui ?

Oh, good find. Yes, it is!

**Pull request merged**